File: crates/bridge/src/agent/paths.rs
//! Managed Qenex home layout under `~/.qenex`.

use std::fs;
use std::path::PathBuf;

pub fn qenex_home() -> PathBuf {
    dirs::home_dir()
        .unwrap_or_else(|| PathBuf::from("."))
        .join(".qenex")
}

pub fn runtime_bun_dir() -> PathBuf {
    qenex_home().join("runtime").join("bun")
}

pub fn runtime_uv_dir() -> PathBuf {
    qenex_home().join("runtime").join("uv")
}

pub fn agents_dir() -> PathBuf {
    qenex_home().join("agents")
}

pub fn agent_version_dir(agent_id: &str, version: &str) -> PathBuf {
    agents_dir().join(agent_id).join(version)
}

pub fn installed_db_path() -> PathBuf {
    qenex_home().join("installed.json")
}

pub fn registry_cache_path() -> PathBuf {
    qenex_home().join("registry-cache.json")
}

pub fn ensure_qenex_dirs() -> Result<(), String> {
    for dir in [
        qenex_home(),
        runtime_bun_dir(),
        runtime_uv_dir(),
        agents_dir(),
    ] {
        fs::create_dir_all(&dir).map_err(|e| format!("create {}: {e}", dir.display()))?;
    }
    Ok(())
}

File: apps/bridge/src/agent/paths.ts
/** Managed Qenex home layout under `~/.qenex`. */
import { existsSync, mkdirSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";

export function qenexHome(): string {
  return join(homedir(), ".qenex");
}

export function runtimeBunDir(): string {
  return join(qenexHome(), "runtime", "bun");
}

export function runtimeUvDir(): string {
  return join(qenexHome(), "runtime", "uv");
}

export function managedBunBin(): string {
  const exe = process.platform === "win32" ? "bun.exe" : "bun";
  return join(runtimeBunDir(), "bin", exe);
}

/** Prefer the interpreter that spawned this Bridge, then managed runtime, then PATH. */
export function resolveBunExecutable(): string {
  const exec = process.execPath;
  if (typeof exec === "string" && /bun/i.test(exec) && existsSync(exec)) {
    return exec;
  }
  const managed = managedBunBin();
  if (existsSync(managed)) return managed;
  return Bun.which("bun") ?? "bun";
}

export function agentsDir(): string {
  return join(qenexHome(), "agents");
}

export function hostsDir(): string {
  return join(qenexHome(), "hosts");
}

export function agentVersionDir(agentId: string, version: string): string {
  return join(agentsDir(), agentId, version);
}

export function installedDbPath(): string {
  return join(qenexHome(), "installed.json");
}

export function registryCachePath(): string {
  return join(qenexHome(), "registry-cache.json");
}

export function ensureQenexDirs(): void {
  for (const dir of [
    qenexHome(),
    runtimeBunDir(),
    runtimeUvDir(),
    agentsDir(),
    hostsDir(),
  ]) {
    mkdirSync(dir, { recursive: true });
  }
}
